File: ios/LaneShadow/Views/Molecules/Banner.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Banner Variant

/**
 * Banner variant enum
 *
 * Defines the four semantic variants for the Banner component.
 * Each variant maps to a theme color for tinting and borders.
 */
public enum LSBannerVariant: Sendable {
    case info
    case success
    case warning
    case error

    /// Returns the theme color associated with this variant
    var color: ThemeColors.KeyPath<ColorSet> {
        switch self {
        case .info: return \.info
        case .success: return \.success
        case .warning: return \.warning
        case .error: return \.danger
        }
    }
}

// MARK: - Banner Component

/**
 * Banner molecule component
 *
 * Displays contextual messages with optional icon, action button, and dismiss control.
 * Following React Native component from react-native/components/ui/molecules/banner.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: variant color at 10% opacity
 *   - Border: variant color at 30% opacity (1pt stroke)
 *   - Text: `theme.colors.onSurface.default`
 *   - Icon/Dismiss: variant color
 * - Layout:
 *   - Corner radius: 8pt
 *   - Padding: horizontal 12pt, vertical 8pt
 *   - Spacing: 8pt between elements
 * - Typography:
 *   - Message: 14pt regular
 *
 * ## Parameters
 * - message: Banner message text
 * - icon: Optional left icon (any SwiftUI View)
 * - action: Optional action button on right (any SwiftUI View)
 * - variant: Color variant (info, success, warning, error)
 * - dismissible: Whether to show close button (default: false)
 * - onDismiss: Optional callback when dismiss button tapped
 */
public struct LSBanner: View {
    @Environment(\.theme) private var theme

    private let message: String
    private let icon: AnyView?
    private let action: AnyView?
    private let variant: LSBannerVariant
    private let dismissible: Bool
    private let onDismiss: (() -> Void)?

    public init(
        message: String,
        icon: AnyView? = nil,
        action: AnyView? = nil,
        variant: LSBannerVariant = .info,
        dismissible: Bool = false,
        onDismiss: (() -> Void)? = nil
    ) {
        self.message = message
        self.icon = icon
        self.action = action
        self.variant = variant
        self.dismissible = dismissible
        self.onDismiss = onDismiss
    }

    // MARK: - Body

    public var body: some View {
        HStack(alignment: .top, spacing: 8) {
            // Optional left icon
            if let icon {
                icon
                    .frame(width: 20, height: 20)
            }

            // Message text area
            Text(message)
                .font(.system(size: 14, weight: .regular))
                .foregroundStyle(theme.colors.onSurface.default)
                .frame(maxWidth: .infinity, alignment: .leading)

            // Optional action button
            if let action {
                action
            }

            // Dismiss button
            if dismissible {
                Button(action: {
                    onDismiss?()
                }) {
                    Image(systemName: "xmark")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(variantColor)
                        .frame(width: 20, height: 20)
                }
                .buttonStyle(PlainButtonStyle())
                .accessibilityLabel("Dismiss")
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(variantColor.opacity(0.1))
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(variantColor.opacity(0.3), lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 8))
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Banner: \(message)")
    }

    // MARK: - Computed Properties

    /// Returns the theme color for the current variant
    private var variantColor: Color {
        theme.colors[keyPath: variant.color]
    }
}

// MARK: - Preview

#Preview("Banner - Info") {
    VStack(spacing: 16) {
        LSBanner(message: "This is an informational banner")
        LSBanner(
            message: "Banner with icon",
            icon: AnyView(
                Image(systemName: "info.circle.fill")
                    .foregroundStyle(Color.blue)
            )
        )
        LSBanner(
            message: "Dismissible banner with message",
            dismissible: true,
            onDismiss: {
                print("Dismissed")
            }
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("Banner - Success") {
    LSBanner(
        message: "Your changes have been saved successfully",
        variant: .success,
        icon: AnyView(
            Image(systemName: "checkmark.circle.fill")
                .foregroundStyle(Color.green)
        )
    )
    .padding()
    .laneShadowTheme()
}

#Preview("Banner - Warning") {
    LSBanner(
        message: "Your session will expire in 5 minutes",
        variant: .warning,
        icon: AnyView(
            Image(systemName: "exclamationmark.triangle.fill")
                .foregroundStyle(Color.orange)
        ),
        action: AnyView(
            Button("Renew") {
                print("Renew tapped")
            }
            .font(.system(size: 14, weight: .semibold))
            .foregroundStyle(Color.orange)
        )
    )
    .padding()
    .laneShadowTheme()
}

#Preview("Banner - Error") {
    LSBanner(
        message: "Failed to load content. Please try again.",
        variant: .error,
        icon: AnyView(
            Image(systemName: "xmark.circle.fill")
                .foregroundStyle(Color.red)
        ),
        dismissible: true,
        onDismiss: {
            print("Dismissed error")
        }
    )
    .padding()
    .laneShadowTheme()
}

#Preview("Banner - Complete") {
    VStack(spacing: 16) {
        LSBanner(
            message: "New ride available in your area",
            variant: .info,
            icon: AnyView(
                Image(systemName: "bell.fill")
                    .foregroundStyle(Color.blue)
            ),
            action: AnyView(
                Button("View") {
                    print("View tapped")
                }
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(Color.blue)
            )
        )

        LSBanner(
            message: "Profile updated successfully",
            variant: .success,
            icon: AnyView(
                Image(systemName: "checkmark.circle.fill")
                    .foregroundStyle(Color.green)
            ),
            dismissible: true
        )

        LSBanner(
            message: "Connection lost. Retrying...",
            variant: .warning,
            icon: AnyView(
                Image(systemName: "wifi.exclamationmark")
                    .foregroundStyle(Color.orange)
            )
        )

        LSBanner(
            message: "Unable to complete request",
            variant: .error,
            icon: AnyView(
                Image(systemName: "xmark.octagon.fill")
                    .foregroundStyle(Color.red)
            ),
            action: AnyView(
                Button("Retry") {
                    print("Retry tapped")
                }
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(Color.red)
            ),
            dismissible: true
        )
    }
    .padding()
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/Banner.kt
package com.laneshadow.ui.components.molecules

import com.laneshadow.ui.atoms.Glyphs

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Banner variant types
 *
 * Following RN wrapper API from react-native/components/ui/banner.tsx
 */
enum class BannerVariant {
    Info,
    Success,
    Warning,
    Error,
}

/**
 * Banner molecule component
 *
 * Informational banner with optional icon, message, and action button.
 * Following React Native wrapper patterns from react-native/components/ui/banner.tsx
 *
 * @param message Banner message text
 * @param icon Optional left icon composable
 * @param action Optional action button composable on right
 * @param variant Color variant (Info, Success, Warning, Error)
 * @param dismissible Whether to show close button
 * @param onDismiss Callback when dismiss button is clicked
 * @param modifier Modifier for the banner container
 * @param contentDescription Accessibility description for screen readers
 */
@Composable
fun Banner(
    message: String,
    icon: @Composable (() -> Unit)? = null,
    action: @Composable (() -> Unit)? = null,
    variant: BannerVariant = BannerVariant.Info,
    dismissible: Boolean = false,
    onDismiss: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Get variant color
    val variantColor: Color = when (variant) {
        BannerVariant.Info -> theme.colors.info.default
        BannerVariant.Success -> theme.colors.success.default
        BannerVariant.Warning -> theme.colors.warning.default
        BannerVariant.Error -> theme.colors.danger.default
    }

    // Background color with low alpha (0.1f = 10% opacity)
    val backgroundColor = variantColor.copy(alpha = 0.1f)

    // Border color with medium alpha (0.3f = 30% opacity)
    val borderColor = variantColor.copy(alpha = 0.3f)

    // Text color from theme
    val textColor = theme.colors.onSurface.default

    // Build semantics
    val bannerModifier = modifier.semantics {
        this.contentDescription = contentDescription ?: message
    }

    Surface(
        modifier = bannerModifier,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.md),
        color = backgroundColor,
        border = BorderStroke(1.dp, borderColor),
    ) {
        Row(
            modifier = Modifier
                .padding(horizontal = 12.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Optional left icon
            if (icon != null) {
                Row(
                    modifier = Modifier,
                    horizontalArrangement = Arrangement.Center,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    icon()
                }
            }

            // Message text (takes remaining space)
            Row(
                modifier = Modifier.weight(1f),
                horizontalArrangement = Arrangement.Start,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = message,
                    style = androidx.compose.ui.text.TextStyle(
                        fontSize = 14.sp,
                        fontWeight = androidx.compose.ui.text.font.FontWeight.Normal,
                    ),
                    color = textColor,
                )
            }

            // Optional action button
            if (action != null) {
                Row(
                    modifier = Modifier,
                    horizontalArrangement = Arrangement.End,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    action()
                }
            }

            // Dismiss button (if dismissible)
            if (dismissible) {
                IconButton(
                    onClick = { onDismiss?.invoke() },
                    modifier = Modifier.size(24.dp),
                ) {
                    Icon(
                        imageVector = Glyphs.Default.Close,
                        contentDescription = "Dismiss",
                        tint = textColor,
                    )
                }
            }
        }
    }
}
